### DISCUSSION TITLE
Viem Type Inference ABI not being found

### DISCUSSION THREAD

**Discussion by radulff** [wevm/viem#1027] [Question] [answered]
I have a situation where I believe I followed along the documentation and online information to  make it work, but I haven't managed to do so yet.

I have migrated from Viem 0.8.x towards 1.4.1. This is the function I try to read from my ABI:

```typescript
{
		inputs: [],
		name: 'getTreasuryAndPhilantrophyBalanceInUSD',
		outputs: [
			{
				internalType: 'uint256',
				name: 'treasuryBalance',
				type: 'uint256',
			},
			{
				internalType: 'uint256',
				name: 'philantrophyBalance',
				type: 'uint256',
			},
		],
		stateMutability: 'view',
		type: 'function',
},
```

I export my ABI as a const, in a TS file:

```typescript
import { Abi } from 'viem';

export const abiTest: Abi = [...] as const;
```

This is my ReadContract implementation:

```typescript
const resultRead = await client.readContract({
		abi: abiCfaPeriphery,
		address: '...' as Address,
		functionName: 'getTreasuryAndPhilantrophyBalanceInUSD',
	});

	console.log('Result Read: ', resultRead);
```

My console log appears as:

```shell
Result Read:  [ 902828817075004553524n, 0n ]
```

Before, I used to get:
```shell
Result Read:  [ 902828817075004553524n, 0n, treasuryBalance:xxxx, philantrophyBalance:xxxxx ]
```

So in my TypeScript implementation I could do:

```typescript
console.log(resultRead.treasuryBalance);
console.log(resultRead.philantrophyBalance);
```

But somehow I have lost this ability, which has broken all my implementation. Is there's something I'm missing?

I have read all the information here:

[Viem Type Inference Explanation](https://viem.sh/docs/typescript.html#type-inference)

[TypeScript Const assertions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions)

Thanks

**Comment by tmm** [comment]
This line looks incorrect.

```ts
export const abiTest: Abi = [...] as const;
```

You shouldn't declare as `Abi`. If you want to check the array to see if it conforms to `Abi`, you should use the [satisfies](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html) operator.

```ts
export const abiTest = [...] as const satisfies Abi;
```

**Reply by radulff** [reply]
HI @tmm, I have changed and tried, but still I can't manage to get the values indexed in the array. What else could it be?

**Reply by tmm** [reply, marked as answer]
Misread your post earlier. I don't think Viem ever supported hybrid-array object return types.

The issue is explained here: https://viem.sh/docs/faq.html#why-is-a-contract-function-return-type-returning-an-array-instead-of-an-object

You can see the difference in this [TypeScript Playground](https://www.typescriptlang.org/play?ts=5.0.4#code/JYWwDg9gTgLgBAbzgYygUwIYzQBQK4BGANsMgMIloB2MANHABYwxj0wCeYacAggcHAC+cAGZQIIOAHIAbsDQgpAKCXIIVAM7wM-OAF44AbSVxEJ03GBUweGBoBcRgLq1zpqhhBpHUgOZoYABV0DA08KHYeKgATHAZgIgwacTAGdgAhDESqZDQASSoAVQBlABEpVws4CFsbO0djKtMENyarbCgPIkDOb2k8doAmAFYANgrWqo8vHxgQsIjM7NyJposOLh8BmhHxyqbBfaqWtdN2tE6sns3+obHV0+m+qVSEpLmIVIyspJWjpo2z22MF2Dyqh1aLlaWiwaAAsrYdAlgBwfHI0AB3MFwQE+ER4HIwYDqB4Q5qtKx1BzOf5PHz+ILzcKRGJxN7JT5pJa-fJFMoAfX5MDwYCIaGxNRYtmpjSaJ1OuOkwtF4v+VTU4HU1HqRkmcr1a3Ol26vS2dz2BqadKVTMWPxyqstVUVUmBoLVBw9xydFiNXWuQPN2NOcGtL3i2Q+X25DuDCtNtx29y9FjJpyhazTpgzFhh2ARMCRJFR0nRWP+LvxhOJVFJriccFCKHUWjgGiwwA0InkGl4-BUak08GQlBo+hQIWw+GIpAo8hoAAokHMkhpILBHEwWAuAJRCHcDlvwdC9gwYDEYFEoUcwAB0IWiZHUK+QMCX5iR+ww0WiJ+pUgABgADweKtXxrAA5TxngZYJMAWFlYgjd4Ui5e1cgKEpylcQQDxPJQAHoCLgAA9AB+Q8hzgE9BWVMVx3PS9hxve9MEfZ8oAwV931MT8Px-P8fGA0CCXA9QoJmaRYNtRC2UjVDvmWXksNokUxQmXClBooU1LQQjiPIpQgA).
